pex/dist_metadata.py

```diff
@@ -69,183 +71,317 @@ class MetadataNotFoundError(MetadataError):
-                    return _parse_message(fp.read())
-            except KeyError as e:
-                pex_warnings.warn(
-                    "Source distribution {} did not have the expected metadata file {}: {}".format(
-                        sdist_path, pkg_info_path, e
-                    )
-                )
-                return None
 
-    return None
+class MetadataType(Enum["MetadataType.Value"]):
+    class Value(Enum.Value):
+        def load_metadata(
+            self,
+            location,  # type: Text
+            project_name=None,  # type: Optional[ProjectName]
+            rescan=False,  # type: bool
+        ):
+            # type: (...) -> Optional[MetadataFiles]
+            return load_metadata(
+                location, project_name=project_name, restrict_types_to=(self,), rescan=rescan
+            )
+
+    DIST_INFO = Value(".dist-info")
+    EGG_INFO = Value(".egg-info")
+    PKG_INFO = Value("PKG-INFO")
 
 
 @attr.s(frozen=True)
-class DistMetadataFile(object):
-    project_name = attr.ib()  # type: ProjectName
-    version = attr.ib()  # type: Version
-    path = attr.ib()  # type: Text
+class MetadataKey(object):
+    metadata_type = attr.ib()  # type: MetadataType.Value
+    location = attr.ib()  # type: Text
 
 
-def find_dist_info_files(
-    filename,  # type: Text
-    listing,  # type: Iterable[Text]
+def _find_installed_metadata_files(
+    location,  # type: Text
+    metadata_type,  # type: MetadataType.Value
+    metadata_dir_glob,  # type: str
+    metadata_file_name,  # type: Text
 ):
-    # type: (...) -> Iterator[DistMetadataFile]
-
-    # N.B. We know the captured project_name and version will not contain `-` even though PEP-503
-    # allows for them in project names and PEP-440 allows for them in versions in some
-    # circumstances. This is since we're limiting ourselves to the products of installs by our
-    # vendored versions of wheel and pip which turn `-` into `_` as explained in `ProjectName` and
-    # `Version` docs.
-    dist_info_metadata_pattern = "^{}$".format(
-        os.path.join(r"(?P<project_name>.+)-(?P<version>.+)\.dist-info", re.escape(filename))
-    )
-    wheel_metadata_re = re.compile(dist_info_metadata_pattern)
-    for item in listing:
-        match = wheel_metadata_re.match(item)
-        if match:
-            yield DistMetadataFile(
-                project_name=ProjectName(match.group("project_name")),
-                version=Version(match.group("version")),
-                path=item,
+    # type: (...) -> Iterator[MetadataFiles]
+    metadata_files = glob.glob(os.path.join(location, metadata_dir_glob, metadata_file_name))
+    for path in metadata_files:
+        with open(path, "rb") as fp:
+            metadata = _parse_message(fp.read())
+            project_name_and_version = ProjectNameAndVersion.from_parsed_pkg_info(
+                source=path, pkg_info=metadata
+            )
+            metadata_file_name = os.path.basename(path)
```

Comment:
Overwriting `metadata_file_name` makes this function a little confusing to grok. Wouldn't this always be identical to the passed-in `metadata_file_name` arg anyway? Assuming we trust that `glob.glob` does what it says on the tin, why re-assign this to the value it already holds?

Reply:
Agreed. That appears to be weird refactor leftover. I'll fix.

Reply:
is the original `metadata_file_name` guaranteed to not include any wildcards? otherwise this is not always the same.

Reply:
Yeah, the names come from a limited set of <10 and Pex, in particular, cares about 5 IIRC.

Reply:
Fixed.